pkg/operator/watchdog/cmd.go

```diff
@@ -0,0 +1,316 @@
+package watchdog
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/server"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+
+	"github.com/openshift/library-go/pkg/config/client"
+	"github.com/openshift/library-go/pkg/controller/fileobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/retry"
+)
+
+type FileWatcherOptions struct {
+	// ProcessName is the name of the process we will send SIGTERM
+	ProcessName string
+
+	// Files lists all files we want to monitor for changes
+	Files      []string
+	KubeConfig string
+
+	// Namespace to report events to
+	Namespace string
+	recorder  events.Recorder
+
+	// Interval specifies how aggressive we want to be in file checks
+	Interval time.Duration
+
+	// for unit-test to mock getting the process PID
+	findPidByNameFn func(name string) (int, bool, error)
+
+	// for unit-test to mock sending UNIX signals
+	handleSignalFn func(pid int) error
+
+	// for unit-test to mock prefixing files (/proc/PID/root)
+	addProcPrefixToFilesFn func([]string, int) []string
+
+	// lastUsedPid is used to prevent sending multiple SIGTERM's to the same process
+	lastUsedPid int
+}
+
+func NewFileWatcherOptions() *FileWatcherOptions {
+	return &FileWatcherOptions{
+		findPidByNameFn:        FindPidByName,
+		addProcPrefixToFilesFn: addProcPrefixToFiles,
+		handleSignalFn: func(pid int) error {
+			return syscall.Kill(pid, syscall.SIGTERM)
+		},
+	}
+}
+
+// NewFileWatcherWatchdog return the file watcher watchdog command.
+func NewFileWatcherWatchdog() *cobra.Command {
+	o := NewFileWatcherOptions()
+
+	cmd := &cobra.Command{
+		Use:   "file-watcher-watchdog",
+		Short: "Watch files on the disk and kill the specified process on change",
+		Run: func(cmd *cobra.Command, args []string) {
+			klog.V(1).Info(cmd.Flags())
+			klog.V(1).Info(spew.Sdump(o))
+
+			// Handle shutdown
+			termHandler := server.SetupSignalHandler()
+			ctx, shutdown := context.WithCancel(context.TODO())
+			go func() {
+				defer shutdown()
+				<-termHandler
+			}()
+
+			if err := o.Complete(); err != nil {
+				klog.Fatal(err)
+			}
+			if err := o.Validate(); err != nil {
+				klog.Fatal(err)
+			}
+
+			if err := o.Run(ctx); err != nil {
+				klog.Fatal(err)
+			}
+		},
+	}
+
+	o.AddFlags(cmd.Flags())
+
+	return cmd
+}
+
+func (o *FileWatcherOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&o.ProcessName, "process-name", "", "name of the process to send signal to on file change (eg. 'hyperkube').")
```

Comment:
which signal? And only once?

Reply:
do we want to give user choice? it is TERM

Reply:
no, but document :)

SIGTERM is pretty aggressive. I.e. not graceful shutdown. Intentional?

Reply:
normally: SIGINT, wait seconds, SIGTERM. A configurable delay would be good (again, graceful shutdown).